libfwupdplugin/fu-security-attrs.c

```diff
@@ -329,44 +329,67 @@ fu_security_attrs_sort_cb(gconstpointer item1, gconstpointer item2)
 void
 fu_security_attrs_depsolve(FuSecurityAttrs *self)
 {
-	g_autoptr(GHashTable) attrs_by_id = NULL;
-
 	g_return_if_fail(FU_IS_SECURITY_ATTRS(self));
 
-	/* make hash of ID -> object */
-	attrs_by_id = g_hash_table_new(g_str_hash, g_str_equal);
-	for (guint i = 0; i < self->attrs->len; i++) {
-		FwupdSecurityAttr *attr = g_ptr_array_index(self->attrs, i);
-		g_hash_table_insert(attrs_by_id,
-				    (gpointer)fwupd_security_attr_get_appstream_id(attr),
-				    (gpointer)attr);
-	}
-
 	/* set flat where required */
 	for (guint i = 0; i < self->attrs->len; i++) {
 		FwupdSecurityAttr *attr = g_ptr_array_index(self->attrs, i);
 		GPtrArray *obsoletes = fwupd_security_attr_get_obsoletes(attr);
-		for (guint j = 0; j < obsoletes->len; j++) {
-			const gchar *obsolete = g_ptr_array_index(obsoletes, j);
-			FwupdSecurityAttr *attr_tmp = g_hash_table_lookup(attrs_by_id, obsolete);
-
-			/* by AppStream ID */
-			if (attr_tmp != NULL) {
-				g_debug("security attr %s obsoleted by %s",
-					obsolete,
-					fwupd_security_attr_get_appstream_id(attr_tmp));
-				fwupd_security_attr_add_flag(attr_tmp,
-							     FWUPD_SECURITY_ATTR_FLAG_OBSOLETED);
+
+		for (guint j = 0; j < self->attrs->len; j++) {
+			FwupdSecurityAttr *attr_tmp = g_ptr_array_index(self->attrs, j);
+
+			/* skip self */
+			if (g_strcmp0(fwupd_security_attr_get_plugin(attr),
+				      fwupd_security_attr_get_plugin(attr_tmp)) == 0 &&
+			    g_strcmp0(fwupd_security_attr_get_appstream_id(attr),
+				      fwupd_security_attr_get_appstream_id(attr_tmp)) == 0)
+				continue;
+
+			/* look for duplicate (negative) attributes when obsolete not explicitly set
+			 */
```

Comment:
`add duplicate (negative) attributes when obsolete not explicitly set` might be more accurate

Reply:
fixed